### PR TITLE
RBD: improvments to existing code to better align with volumegroup

### PR DESCRIPTION
**Pull request by Madhu-1**
This PR contains the commits taken out of https://github.com/ceph/ceph-csi/pull/4739 to make the other PR smaller. This contains lot of improvement which is going to help for future implementation and this also takes care of handling the flattening before adding the image to the group.

### CODE REVIEW

File: internal/rbd/mirror.go [hunk 1]
```diff
@@ -61,7 +61,7 @@ func (rv *rbdVolume) HandleParentImageExistence(
 
 	parent, err := rv.getParent()
 	if err != nil {
-		return err
+		return fmt.Errorf("failed to get parent image %s: %w", rv, err)
```
Comment (Rakshith-R) [inline internal/rbd/mirror.go]:
```suggestion
		return fmt.Errorf("failed to get parent of image %s: %w", rv, err)
```

File: internal/csi-addons/rbd/volumegroup.go [hunk 2]
```diff
@@ -341,17 +341,40 @@ func (vs *VolumeGroupServer) ModifyVolumeGroupMembership(
...
+		if err != nil {
 			return nil, status.Errorf(
 				codes.NotFound,
 				"failed to find a volume with CSI ID %q: %v",
 				id,
 				err)
 		}
 
+		//nolint:gocritic // need to call .Destroy() for all volumes
+		defer vol.Destroy(ctx)
```
Comment (Rakshith-R) [inline internal/csi-addons/rbd/volumegroup.go]: This deferred function is called when the outer function is exiting right ? 
not when the loop ends or all iterations are complete ? 
(I maybe mistaken)

File: internal/rbd/group/volume_group.go [hunk 3]
```diff
@@ -151,18 +151,17 @@ func GetVolumeGroup(
 	return vg, nil
 }
 
-// String makes it easy to include the volumeGroup object in log and error
-// messages.
+// String returns the image-spec (pool/{namespace}/{name}) format of the group.
 func (vg *volumeGroup) String() string {
```
Comment (Rakshith-R) [inline internal/rbd/group/volume_group.go:155]: tagging @nixpanic to verify is vg.id is not used anywhere from the Stringer call.

**Review by Rakshith-R — COMMENTED** [review]
tiny nits

File: internal/csi-addons/rbd/volumegroup.go (hunk 2 above)
Comment (Madhu-1) [reply internal/csi-addons/rbd/volumegroup.go]: Yes, you are correct, this gets stacked up as it's called in for loop. removed it

File: internal/csi-addons/rbd/volumegroup.go [hunk 4]
```diff
@@ -117,6 +117,19 @@ func (vs *VolumeGroupServer) CreateVolumeGroup(
...
+	flattenMode, err := getFlattenMode(ctx, req.GetParameters())
+	if err != nil {
+		return nil, err
+	}
+	// Flatten the image if the flatten mode is set to FlattenModeForce
+	// before adding it to the volume group.
+	for _, v := range volumes {
+		err = v.HandleParentImageExistence(ctx, flattenMode)
+		if err != nil {
+			return nil, getGRPCError(err)
```
Comment (nixpanic) [inline internal/csi-addons/rbd/volumegroup.go:135]: include a bit more details in the error maybe? Mention the VolumeGroup and Volume?

File: internal/csi-addons/rbd/volumegroup.go (hunk 2 above)
Comment (nixpanic) [reply internal/csi-addons/rbd/volumegroup.go]: an other option, maybe cleaner?

```go
volumes := make([]types.Volume, len(toAdd))
defer func() {
    // need to call .Destroy() for all volumes
    for _, vol := range volume {
        vol.Destroy(ctx)
    }
}()

for i, id := range toAdd {
    ...
```

File: internal/csi-addons/rbd/replication.go [hunk 5]
```diff
@@ -284,9 +284,9 @@ func (rs *ReplicationServer) EnableVolumeReplication(ctx context.Context,
 	if err != nil {
 		switch {
 		case errors.Is(err, corerbd.ErrImageNotFound):
-			err = status.Errorf(codes.NotFound, "volume %s not found", volumeID)
+			err = status.Errorf(codes.NotFound, err.Error())
```
Comment (nixpanic) [inline internal/csi-addons/rbd/replication.go]: code duplication for `NotFound` error cases, consider using `fallthrough`:

```go
switch {
    case errors.Is(err, corerbd.ErrImageNotFound):
        fallthrough
    case errors.Is(err, util.ErrPoolNotFound):
        err = status.Errorf(codes.NotFound, err.Error())
...
```

this code block is repeated really a lot, it is probably worth it to put it in a helper function.

File: internal/rbd/group/volume_group.go (hunk 3 above)
Comment (nixpanic) [reply internal/rbd/group/volume_group.go:155]: it shouldn't, `vg.GetID()` is the preferred way, or in some cases `vg.id` is safe.

File: internal/csi-common/utils.go [hunk 6]
```diff
@@ -131,6 +131,9 @@ func GetIDFromReplication(req interface{}) string {
 		if reqID == "" {
 			reqID = r.GetVolumeId() //nolint:nolintlint,staticcheck // req.VolumeId is deprecated
 		}
+		if reqID == "" {
+			reqID = src.GetVolumegroup().GetVolumeGroupId()
```
Comment (nixpanic) [inline internal/csi-common/utils.go]: can `GetVolumegroup()` return nil?

**Review by nixpanic — CHANGES_REQUESTED** [review]
few minor things that should be improved. Thanks for splitting these commits out of the larger one!

File: internal/csi-addons/rbd/volumegroup.go (hunk 2 above)
Comment (Madhu-1) [reply internal/csi-addons/rbd/volumegroup.go]: i did think and reverted back, just before you reviewed the PR, will bring it back

File: internal/csi-addons/rbd/volumegroup.go (hunk 4 above)
Comment (Madhu-1) [reply internal/csi-addons/rbd/volumegroup.go:135]: added volumegroup, volume details are already returned by HandleParentImageExistence

File: internal/csi-common/utils.go (hunk 6 above)
Comment (Madhu-1) [reply internal/csi-common/utils.go]: yes, missed this one, fixed now, added a test for it as well

File: internal/rbd/manager.go [hunk 7]
```diff
@@ -124,7 +124,18 @@ func (mgr *rbdManager) GetVolumeByID(ctx context.Context, id string) (types.Volu
...
 	volume, err := GenVolFromVolID(ctx, id, creds, mgr.secrets)
 	if err != nil {
-		return nil, fmt.Errorf("failed to get volume from id %q: %w", id, err)
+		switch {
+		case errors.Is(err, ErrImageNotFound):
+			err = fmt.Errorf("volume %s not found: %w", id, err)
+
+			return nil, err
+		case errors.Is(err, util.ErrPoolNotFound):
+			err = fmt.Errorf("pool %s not found for %s: %w", volume.Pool, id, err)
```
Comment (nixpanic) [inline internal/rbd/manager.go:133]: `volume.Pool` seems weird, the `Volume` interface does not have a `Pool` attribute.

On a second look, the `GenVolFromVolID()` function returns an `rbdVolume`, so it should work.

**Review by nixpanic — APPROVED** [review]

**Review by Rakshith-R — APPROVED** [review]

**Comment by Rakshith-R** [comment]
@mergifyio queue

**Comment by nixpanic** [comment]
@mergifyio requeue

**Pull request merged**